conda/resolve.py

```diff
@@ -327,29 +327,67 @@ def filter_group(matches):
 
         # Determine all valid packages in the dependency graph
         reduced_index = {}
-        slist = list(specs)
+        explicit_spec_list = list(specs)
         for fstr in features:
             dist = Dist(fstr + '@')
             reduced_index[dist] = self.index[dist]
-        while slist:
-            this_spec = slist.pop()
-            for dist in self.find_matches(this_spec):
-                if reduced_index.get(dist) is None and self.valid(dist, filter):
-                    reduced_index[dist] = self.index[dist]
-                    for ms in self.ms_depends(dist):
-                        # We do not pull packages into the reduced index due
-                        # to a track_features dependency. Remember, a feature
-                        # specifies a "soft" dependency: it must be in the
-                        # environment, but it is not _pulled_ in. The SAT
-                        # logic doesn't do a perfect job of capturing this
-                        # behavior, but keeping these packags out of the
-                        # reduced index helps. Of course, if _another_
-                        # package pulls it in by dependency, that's fine.
-                        if 'track_features' not in ms:
-                            slist.append(ms)
+        seen_top_level_pkgs = set()
+        for explicit_spec in explicit_spec_list:
+            explicit_packages = set(self.find_matches(explicit_spec)) - seen_top_level_pkgs
+            for pkg in explicit_packages:
+                # what we have seen is only relevant within the context of a single package
+                #    that is picked up because of an explicit spec.  We don't want the
+                #    broadening check to apply across packages at the explicit level; only
+                #    at the level of deps below that explicit package.
+                seen_pkgs = {pkg}
+                seen_specs = set()
+                seen_top_level_pkgs.add(pkg)
+                if pkg not in reduced_index and self.valid(pkg, filter):
```

Comment:
An indentation could be saved here using:
```
if pkg in reduced_index or not self.valid(pkg, filter):
    continue
```

Reply:
good one.  it is very deeply indented, and this helps.